utils/diffing.ml

```diff
@@ -38,337 +38,343 @@ type ('left, 'right, 'eq, 'diff) change =
   | Keep of 'left * 'right * 'eq
   | Change of 'left * 'right * 'diff
 
-type ('l, 'r, 'eq, 'diff) patch = ('l, 'r, 'eq, 'diff) change list
+type ('left, 'right, 'eq, 'diff) patch = ('left, 'right, 'eq, 'diff) change list
 
-let map f g = function
-  | Delete x -> Delete (f x)
-  | Insert x -> Insert (g x)
-  | Keep (x,y,k) -> Keep (f x, g y, k)
-  | Change (x,y,k) -> Change (f x, g y, k)
+module type Diffable = sig
+  type left
+  type right
 
-type ('st,'left,'right) full_state = {
-  line: 'left array;
-  column: 'right array;
-  state: 'st
-}
+  type eq
+  type diff
 
-(* The matrix supporting our dynamic programming implementation.
+  (** The type of potential changes on a list. *)
+  type nonrec change = (left, right, eq, diff) change
 
-   Each cell contains:
-   - The diff and its weight
-   - The state computed so far
-   - The lists, potentially extended locally.
+  (** A patch is an ordered list of changes. *)
+  type patch = change list
 
-   The matrix can also be reshaped.
-*)
-module Matrix : sig
+  val weight : change -> int
+
+  type state
 
-  type shape = { l : int ; c : int }
+  (** [test st xl xr] tests if the elements [xl] and [xr] are
+      compatible ([Ok]) or not ([Error]). *)
+  val test : state -> left -> right -> (eq, diff) result
 
-  type ('state,'left,'right,'eq,'diff) t
+  (** [update change state] return the new state after applying a change.
 
-  val make : shape -> ('st,'l,'r,'e,'d) t
-  val reshape : shape -> ('st,'l,'r,'e,'d) t -> ('st,'l,'r,'e,'d) t
+     We support variadic diffing: dynamically extending the lists to
+     diff depending on the placement choices for a prefix of the
+     input. This is done by returning (optional) extensions for the
+     left or right input array. *)
+  val update : change -> state -> state * left array option * right array option
```

Comment:
The option type is splitting the "no extension" value into `None | Some [||]`. If the aim is to explicit the fact that extensions are not mandatory, I would rather kept the not variadic variant (and ideally keep the limitation that extension are only allowed on one side).

Reply:
I was going to make the same comment: the previous (slightly convoluted) type was there on purpose, we want to enforce that users are only extending on one side only, and that it doesn't change during execution.

Reply:
The change of type is indeed blurring the distinction between "I statically know that we will never extend along this direction" and "dynamically I don't want to extend this time". It results in a substantial simplification -- but we could always revert to the previous complex API.
(One thing I can do but haven't yet is get rid of the dynamic `= [||]` check in the diffing code, given that producers now choose None in this situation instead.)

Why is it important to prevent an update function from trying to update both directions? I don't understand the details of the algorithm but the code suggests that extending alternatively in either directions would in fact be fine.

Reply:
Termination is not guaranteed if the client can extends on both side (it could alternate a bit on each side in a loop).

We could just write it in the documentation, but this is a low-usage API, so we might use the inconvenient-but-safe API.

Reply:
Is termination guaranteed if the left update function always add a new line? The column width remains fixed, but I would naively assume that you need to explore the infinitely-growing lines in case they contain a better path?

> it's too easy to do a local change in Includemod_error (for instance), without realizing that an invariant is broken.

What is the invariant we are talking about? If it is "we only add in the same direction", why is it worth enforcing in the code?

Reply:
Yes, it's guaranteed, as long as it's always in the direction, because at some point we stop exploring things (we reach the bottom right corner), regardless if the update function wants to add more things. We only explore new lines if we need them to produce a diff. The termination argument is given in the code (and maybe a bit in the original PR).

In an another word, an infinitely long rectangle is okay because it's still bounded on one side and we ignore the trailing deletions. An infinite square isn't.

So yes, "we only add in the same direction" is the invariant, and it's the right one for termination (and it's one that fits well with the client code too).

Reply:
I don't get it. The code reads as follows:

```ocaml
  (* Filling the matrix

     We fill the whole matrix, as in vanilla Wagner-Fischer.
     At this point, the lists in some states might have been extended.
     If any list have been extended, we need to reshape the matrix
     and repeat the process
  *)
  let compute_matrix state0 =
    let m0 = Matrix.make { l = 0 ; c = 0 } in
    Matrix.set m0 0 0 ~weight:0 ~state:state0 ~diff:None;
    let rec loop m =
      let shape = Matrix.shape m in
      let new_shape = Matrix.real_shape m in
      if new_shape.l > shape.l || new_shape.c > shape.c then
        let m = Matrix.reshape new_shape m in
        for i = 0 to new_shape.l do
          for j = 0 to new_shape.c do
            compute_cell m i j
          done
        done;
        loop m
      else
        m
    in
    loop m0
```

If all calls to the `update` function inside `compute_cell` add a new line, then I would expect the test `new_shape.l > shape.l` to always hold, so that the function loops. What am I missing?

(My understanding is that `real_shape` computes the maximum line height and the maximum column width over all states of the the `shape`-delimited rectangle.)

Reply:
The missing constraint is that extension should only happen in the Keep/Change cases. Then, if extension only happens on one side, the non-extensible side will be exhausted eventually.

Reply:
Hm hm... I can see that you both suffer from Stockhol syndrome with your ugly `update` type definition, but I was never its captive so I still don't see a strong justification for it. (Currently the justification given is approximatively "it gives static guarantees that another dynamic property implies termination, while your version requires two different dynamic properties for termination".)

On the other hand, according to the principle that you should always accept some of the reviewers suggestions, I am going to go ahead and move the Diffable modules inside the function instead of functorizing them over the functor argument. Do we have a deal this way?

Reply:
Or, alternatively, you don't really understand the algorithm and you never had to debug it in anger. :)

Maybe our explanation didn't made that clear enough: The termination *is* enforced statically, there is no dynamic property involved. `update` is only called in the `keep/change` case, which corresponds to a diagonal move in the matrix. If one direction is bound, there can only be a finite number of such diagonal moves, thus a finite number of extensions. The "infinite" rectangle is hypothetical: maybe your update would do it, but the algorithm can never give it such opportunity.

So I'm going to insist, I want my ugly `update` which encodes exactly the right property.

Reply:
> `update` is only called in the `Keep/Change` case, which corresponds to a diagonal move in the matrix.

I don't get it. Here is the code I see in trunk:

```ocaml
let compute_inner_cell ~weight ~test ~update tbl i j =
  let compute_proposition i j diff =
    let* diff = diff in
    let+ localstate = Matrix.state tbl i j in
    weight diff + Matrix.weight tbl i j, (diff, localstate)
  in
  let del =
    let diff = let+ x = Matrix.line tbl (i-1) j in Delete x in
    compute_proposition (i-1) j diff
  in
  let insert =
    let diff = let+ x = Matrix.column tbl i (j-1) in Insert x in
    compute_proposition i (j-1) diff
  in
  let diag =
    let diff =
      let* state = Matrix.state tbl (i-1) (j-1) in
      let* line = Matrix.line tbl (i-1) (j-1) in
      let* column = Matrix.column tbl (i-1) (j-1) in
      match test state.state line column with
      | Ok ok -> Some (Keep (line, column, ok))
      | Error err -> Some (Change (line, column, err))
    in
    compute_proposition (i-1) (j-1) diff
  in
  let*! newweight, (diff, localstate) =
    select_best_proposition [diag;del;insert]
  in
  let state = update diff localstate in
  Matrix.set tbl i j ~weight:newweight ~state ~diff:(Some diff)
```

It looks to me like `update` is called unconditionally, on a `diff` whose category (insert/delete/keep/change) depends on the user-provided `weight` function.

I think that the property you are referring to is the fact that, in the user code in `includemod.ml`, the update function in fact only provides a non-empty array in the Keep/Change case. This is what I call a dynamic property (in particular, there is no enforcement by the type system).

Reply:
Indeed the following code diverges in trunk:

```
let () =
  let li =
    variadic_diff
      ~weight:(function _ -> 0)
      ~test:(fun _ _ _ -> Error ())
      ~update:(With_left_extensions (fun _ () -> (), [|()|]))
      () [|()|] [|()|]
  in
  print_endline (string_of_int (List.length li))
```

Reply:
> Do we have a deal this way?

Not at all.

Exposing variadic diffing as the only interface seems unnecessarily error prone.
(Of course the update type is an implementation detail that can go away).

Similarly, I would rather make the variadic interface as constrained as possible (with one side at a time, which removes a possible mistake at the consumer sode).

And adding those specialized functors seems straigthforward?

Reply:
> And adding those specialized functors seems straigthforward?

Yes, of course. I can also reinstate the ugly `update` (I don't particularly care) if this is a blocking point.

Reply:
The interface could be strengthened to avoid this issue, but your example is in the territory of "clearly malicious code` whereas the current iteration was aimed at relieving the `update` writer from the burden of remembering which side should be extended.